students/luyao_xu/lesson05/mailroom3.py

```diff
@@ -0,0 +1,148 @@
+    return donor
+
+
+def donate_sum(donor):
+    return sum(donor['donation amount'])
+
+
+def donate(donor):
+    # Once a name has been selected, prompt for a donation amount
+    try:
+        donation_amount = float(input('Enter the amount you want to donate: $'))
+        donor = donate_amount(donor, donation_amount)
+        while True:
+            more = input('Is there another gift amount you want to enter? (y/n) ')
+            if more[0].lower() == 'y':
+                donation_amount = float(input('That\'s so nice of you! Enter the amount you want to donate: $'))
+                # Once an amount has been given, add that amount to the donation history of the selected user
+                donor = donate_amount(donor, donation_amount)
+
+            # Use string formatting to compose an email thanking the donor for their generous donation.
+            elif more.lower() == 'n':
+                print(letter(donor))
+                break
+    except ValueError:
+        print("please enter a valid number")
+        donate(donor)
+
+    return donor
+
+
+def create_donor(name):
+    return {'name': name, 'donation amount': []}
+
+
+def send_thank():
+    while True:
+        # If the user types ‘list’, show them a list of the donor names and re-prompt
+
+        user_input = input(
+            'Please enter a full name, type "list" to see the list of donors, type "back" to go back: ')
+        if user_input.lower() == 'list':
+            for name in donations:
+                print(name['name'])
+            continue
+
+        elif user_input.lower() == 'back':
+            # back to the main menu
+            return
+        # If the user types a name in the list, use it.
+        elif user_input.lower() in [d['name'].lower() for d in donations]:
+            for d in donations:
+                if d['name'] == user_input:
+                    donate(d)
+                    # exit for loop on name match
+                    break
+        else:
+            # If the user types a name not in the list, add that name to the data structure and use it
+            d = create_donor(user_input)
+            donations.append(donate(d))
+            break
+
+
+def sort_total_given(k):
+    return sum(k['donation amount'])
+
+
+def report_string(d):
+    result = '{:<26} | {:^11} | {:^8} | {:>12}\n'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+    result += '-' * len(result) + '\n'
+    for i in d:
+        name = i['name']
+        total_donation = sum(i['donation amount'])
+        gift_number = len(i['donation amount'])
+        average_donation = total_donation / gift_number
+
+        result += '{:<27} ${:>11.2f} {:>12} ${:>13.2f}\n'.format(name, total_donation, gift_number, average_donation)
+    return result
+
+
+def create_report():
+    donations.sort(key=sort_total_given, reverse=True)
+    #  print a list of your donors, sorted by total historical donation amount
+    print(report_string(donations))
```

Comment:
I would move the sort code into `report_string` as well, since it's part of that feature.. and rename your functions:

```
def get_report_string():
    """Prepares sorted report and returns as a string"""
    pass


def create_report():
    print(get_report_string())
```